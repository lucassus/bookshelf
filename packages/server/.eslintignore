coverage
dist
node_modules
jest.config.js
src/**/*.generated.ts
