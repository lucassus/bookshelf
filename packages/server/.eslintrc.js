module.exports = {
  extends: "../../.eslintrc.js",
  overrides: [
    {
      files: ["*.test.ts"],
      env: { jest: true },
      extends: ["plugin:jest/recommended"]
    }
  ],
  rules: {
    "@typescript-eslint/no-var-requires": "off",
    "global-require": "off",
    "import/no-extraneous-dependencies": [
      "error",
      { devDependencies: ["**/*.test.js", "src/testUtils/**/*.ts"] }
    ]
  }
};
