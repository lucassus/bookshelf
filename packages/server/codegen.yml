schema:
  ./src/graphql/**/*.graphql:
    noRequire: true

generates:
  ./src/graphql/resolvers-types.generated.ts:
    config:
      useIndexSignature: true
      mappers:
        Author: ../database/entity/Author#Author as AuthorEntity
        Avatar: ../database/entity/Avatar#Avatar as AvatarEntity
        Book: ../database/entity/Book#Book as BookEntity
        BookCopy: ../database/entity/BookCopy#BookCopy as BookCopyEntity
        User: ../database/entity/User#User as UserEntity
      scalars:
        ExternalID: string
    plugins:
      - typescript
      - typescript-resolvers
