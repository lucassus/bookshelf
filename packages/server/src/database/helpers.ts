import { SecureId } from "./SecureId";

type EntityTypes = "Author" | "Book" | "User" | "BookCopy";

export const secureId = new SecureId<EntityTypes>({
  separator: "-"
});
