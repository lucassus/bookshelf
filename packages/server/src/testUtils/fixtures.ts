import { getConnection } from "typeorm";

import { Author } from "../entity/Author";
import { Book } from "../entity/Book";
import { User } from "../entity/User";

async function loadAuthors() {
  const { manager } = getConnection();

  const authors = manager.create(Author, [
    {
      name: "J. K. Rowling",
      photoPath: "/images/book-authors/j-k-rowling.jpg"
    },
    {
      name: "James S. A. Corey",
      photoPath: "/images/book-authors/james-s-a-corey.jpg"
    },
    {
      name: "Andrzej Sapkowski",
      photoPath: "/images/book-authors/andrzej-sapkowski.jpg"
    }
  ]);

  return manager.save(authors);
}

function loadBooks(authors: Author[]) {
  const { manager } = getConnection();

  const books = manager.create(Book, [
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Sorcerer's Stone",
      coverPath: "/images/book-covers/harry1.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Chamber of Secrets",
      coverPath: "/images/book-covers/harry2.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Prisoner of Azkaban",
      coverPath: "/images/book-covers/harry3.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Goblet of Fire",
      coverPath: "/images/book-covers/harry4.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Order of the Phoenix",
      coverPath: "/images/book-covers/harry5.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Half-Blood Prince",
      coverPath: "/images/book-covers/harry6.jpg"
    },
    {
      authorId: authors[0].id,
      title: "Harry Potter and the Deathly Hallows",
      coverPath: "/images/book-covers/harry7.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Leviathan Wakes",
      coverPath: "/images/book-covers/expanse1.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Caliban's War",
      coverPath: "/images/book-covers/expanse2.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Abaddon's Gate",
      coverPath: "/images/book-covers/expanse3.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Cibola Burn",
      coverPath: "/images/book-covers/expanse4.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Nemesis Games",
      coverPath: "/images/book-covers/expanse5.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Babylon's Ashes",
      coverPath: "/images/book-covers/expanse6.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Persepolis Rising",
      coverPath: "/images/book-covers/expanse7.jpg"
    },
    {
      authorId: authors[1].id,
      title: "Tiamat's Wrath",
      coverPath: "/images/book-covers/expanse8.jpg"
    },
    {
      authorId: authors[2].id,
      title: "Blood of Elves",
      coverPath: "/images/book-covers/witcher1.jpg"
    },
    {
      authorId: authors[2].id,
      title: "Time of contempt",
      coverPath: "/images/book-covers/witcher2.jpg"
    },
    {
      authorId: authors[2].id,
      title: "Baptism of fire",
      coverPath: "/images/book-covers/witcher3.jpg"
    },
    {
      authorId: authors[2].id,
      title: "The tower of the swallow",
      coverPath: "/images/book-covers/witcher4.jpg"
    },
    {
      authorId: authors[2].id,
      title: "The lady of the lake",
      coverPath: "/images/book-covers/witcher5.jpg"
    }
  ]);

  return manager.save(books);
}

async function loadUsers() {
  const { manager } = getConnection();

  const users = manager.create(User, [
    {
      name: "Alice",
      email: "alice@example.com",
      avatar: {
        imagePath: "/images/avatars/w13.png",
        color: "yellow"
      }
    },
    {
      name: "Bob",
      email: "bob@example.com",
      avatar: {
        imagePath: "/images/avatars/m10.png",
        color: "green"
      }
    },
    {
      name: "Celine",
      email: "celine@example.com",
      avatar: {
        imagePath: "/images/avatars/w2.png",
        color: "red"
      }
    },
    {
      name: "Dan",
      email: "dan@example.com",
      avatar: {
        imagePath: "/images/avatars/m25.png",
        color: "blue"
      }
    }
  ]);
  await manager.save(users);
}

export const loadFixtures = async () => {
  const authors = await loadAuthors();
  await loadBooks(authors);

  await loadUsers();
};
