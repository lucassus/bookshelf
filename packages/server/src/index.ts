import express from "express";
import path from "path";
import "reflect-metadata";

import { PORT } from "./config";
import { createConnection } from "./database/createConnection";
import { createServer } from "./server";

const startServer = async () => {
  const connection = await createConnection();
  const server = createServer(connection);

  const app = express();
  server.applyMiddleware({ app });

  const distDir = path.join(__dirname, "../../../web/dist");
  app.use(express.static(distDir));
  app.get("/*", (req, res) => {
    res.sendFile(path.join(distDir, "index.html"));
  });

  app.listen({ port: PORT });

  return server;
};

startServer()
  .then((server) => {
    console.log(
      `ðŸš€ Server ready at http://localhost:${PORT}${server.graphqlPath}`
    );
  })
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
