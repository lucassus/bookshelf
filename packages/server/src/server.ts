import { ApolloServer } from "apollo-server-express";
import { Connection } from "typeorm";

import { ASSETS_BASE_URL } from "./config";
import { buildAuthorsLoader } from "./database/authorsLoader";
import { User } from "./database/entity/User";
import { rootSchema } from "./graphql/schema";
import { Context } from "./types";

export const createServer = (connection: Connection) =>
  new ApolloServer({
    schema: rootSchema,
    context: async (): Promise<Context> => {
      const currentUser = await connection.manager.findOne(User, {
        name: "Bob"
      });

      return {
        assetsBaseUrl: ASSETS_BASE_URL,
        connection,
        authorsLoader: buildAuthorsLoader(),
        currentUser
      };
    },
    introspection: true,
    playground: true
  });
