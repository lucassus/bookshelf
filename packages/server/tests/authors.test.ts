import { ApolloServer, gql } from "apollo-server-express";
import { createTestClient } from "apollo-server-testing";

import { secureId } from "../src/database/helpers";
import { loadFixtures } from "../src/fixtures";
import { createServer } from "../src/server";

let server: ApolloServer;

beforeEach(async () => {
  await loadFixtures();
  server = createServer();
});

it("fetches an author", async () => {
  const { query } = createTestClient(server);

  // When
  const res = await query({
    query: gql`
      query GetAuthor($id: ID!) {
        author(id: $id) {
          id
          name
          bio
          books {
            title
          }
        }
      }
    `,
    variables: { id: secureId.toExternal(1, "Author") }
  });

  // Then
  expect(res.data).not.toBeUndefined();
  expect(res.data).toMatchSnapshot();
});

it("fetches authors along with books", async () => {
  // Given
  const { query } = createTestClient(server);

  // When
  const res = await query({
    query: gql`
      query {
        authors {
          name
          books {
            title
          }
        }
      }
    `
  });

  // Then
  expect(res.data).not.toBeUndefined();
  expect(res.data).toMatchSnapshot();
});
