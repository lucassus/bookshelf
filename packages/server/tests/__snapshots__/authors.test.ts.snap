// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches an author 1`] = `
Object {
  "author": Object {
    "bio": "Although she writes under the pen name J.K. Rowling, pronounced like rolling, her name when her first Harry Potter book was published was simply Joanne Rowling. Anticipating that the target audience of young boys might not want to read a book written by a woman, her publishers demanded that she use two initials, rather than her full name. As she had no middle name, she chose K as the second initial of her pen name, from her paternal grandmother Kathleen Ada Bulgen Rowling. She calls herself Jo and has said, \\"No one ever called me 'Joanne' when I was young, unless they were angry.\\" Following her marriage, she has sometimes used the name Joanne Murray when conducting personal business. During the Leveson Inquiry she gave evidence under the name of Joanne Kathleen Rowling. In a 2012 interview, Rowling noted that she no longer cared that people pronounced her name incorrectly.

      Rowling was born to Peter James Rowling, a Rolls-Royce aircraft engineer, and Anne Rowling (née Volant), on 31 July 1965 in Yate, Gloucestershire, England, 10 miles (16 km) northeast of Bristol. Her mother Anne was half-French and half-Scottish. Her parents first met on a train departing from King's Cross Station bound for Arbroath in 1964. They married on 14 March 1965. Her mother's maternal grandfather, Dugald Campbell, was born in Lamlash on the Isle of Arran. Her mother's paternal grandfather, Louis Volant, was awarded the Croix de Guerre for exceptional bravery in defending the village of Courcelles-le-Comte during the First World War.
      
      Rowling's sister Dianne was born at their home when Rowling was 23 months old. The family moved to the nearby village Winterbourne when Rowling was four. She attended St Michael's Primary School, a school founded by abolitionist William Wilberforce and education reformer Hannah More. Her headmaster at St Michael's, Alfred Dunn, has been suggested as the inspiration for the Harry Potter headmaster Albus Dumbledore.
      
      As a child, Rowling often wrote fantasy stories, which she would usually then read to her sister. She recalls that: \\"I can still remember me telling her a story in which she fell down a rabbit hole and was fed strawberries by the rabbit family inside it. Certainly the first story I ever wrote down (when I was five or six) was about a rabbit called Rabbit. He got the measles and was visited by his friends, including a giant bee called Miss Bee.\\" At the age of nine, Rowling moved to Church Cottage in the Gloucestershire village of Tutshill, close to Chepstow, Wales. When she was a young teenager, her great aunt, who Rowling said \\"taught classics and approved of a thirst for knowledge, even of a questionable kind,\\" gave her a very old copy of Jessica Mitford's autobiography, Hons and Rebels. Mitford became Rowling's heroine, and Rowling subsequently read all of her books.
      
      Rowling has said of her teenage years, in an interview with The New Yorker, \\"I wasn’t particularly happy. I think it’s a dreadful time of life.\\" She had a difficult homelife; her mother was ill and she had a difficult relationship with her father (she is no longer on speaking terms with him). She attended secondary school at Wyedean School and College, where her mother had worked as a technician in the science department. Rowling said of her adolescence, \\"Hermione [a bookish, know-it-all Harry Potter character] is loosely based on me. She's a caricature of me when I was eleven, which I'm not particularly proud of.\\" Steve Eddy, who taught Rowling English when she first arrived, remembers her as \\"not exceptional\\" but \\"one of a group of girls who were bright, and quite good at English.\\" Sean Harris, her best friend in the Upper Sixth owned a turquoise Ford Anglia, which she says inspired the one in her books.",
    "books": Array [
      Object {
        "title": "Harry Potter and the Chamber of Secrets",
      },
      Object {
        "title": "Harry Potter and the Deathly Hallows",
      },
      Object {
        "title": "Harry Potter and the Goblet of Fire",
      },
      Object {
        "title": "Harry Potter and the Half-Blood Prince",
      },
      Object {
        "title": "Harry Potter and the Order of the Phoenix",
      },
      Object {
        "title": "Harry Potter and the Prisoner of Azkaban",
      },
      Object {
        "title": "Harry Potter and the Sorcerer's Stone",
      },
    ],
    "id": "MS1BdXRob3I=",
    "name": "J. K. Rowling",
  },
}
`;

exports[`fetches authors along with books 1`] = `
Object {
  "authors": Array [
    Object {
      "books": Array [
        Object {
          "title": "Harry Potter and the Chamber of Secrets",
        },
        Object {
          "title": "Harry Potter and the Deathly Hallows",
        },
        Object {
          "title": "Harry Potter and the Goblet of Fire",
        },
        Object {
          "title": "Harry Potter and the Half-Blood Prince",
        },
        Object {
          "title": "Harry Potter and the Order of the Phoenix",
        },
        Object {
          "title": "Harry Potter and the Prisoner of Azkaban",
        },
        Object {
          "title": "Harry Potter and the Sorcerer's Stone",
        },
      ],
      "name": "J. K. Rowling",
    },
    Object {
      "books": Array [
        Object {
          "title": "Abaddon's Gate",
        },
        Object {
          "title": "Babylon's Ashes",
        },
        Object {
          "title": "Caliban's War",
        },
        Object {
          "title": "Cibola Burn",
        },
        Object {
          "title": "Leviathan Wakes",
        },
        Object {
          "title": "Nemesis Games",
        },
        Object {
          "title": "Persepolis Rising",
        },
        Object {
          "title": "Tiamat's Wrath",
        },
      ],
      "name": "James S. A. Corey",
    },
    Object {
      "books": Array [
        Object {
          "title": "Baptism of fire",
        },
        Object {
          "title": "Blood of Elves",
        },
        Object {
          "title": "The lady of the lake",
        },
        Object {
          "title": "The tower of the swallow",
        },
        Object {
          "title": "Time of contempt",
        },
      ],
      "name": "Andrzej Sapkowski",
    },
  ],
}
`;
